tools/upload_resumable_file.php

```diff
@@ -23,86 +24,119 @@
 $staging_dir = "$root_staging_dir/{$hashed_filename}dir";
 $chunk_filename = "$staging_dir/$hashed_filename.part.$chunk_number";
 
-// handle testChunks request, this allows uploads to be restarted by
+// Handle testChunks (GET) requests, this allows uploads to be restarted by
 // allowing the browser to query for which parts have been uploaded
 // successfully
 if ($_SERVER['REQUEST_METHOD'] === 'GET') {
-    if (file_exists($chunk_filename)) {
-        header("HTTP/1.0 200 Ok");
-        // continue to try to reassemble
+    if (file_exists($chunk_filename) && filesize($chunk_filename) >= $chunk_size) {
+        if ($chunk_number == $total_chunks) {
+            // try to reassemble, see note in reassemble()
+            reassemble();
+        } else {
+            http_response_code(204);
+        }
     } else {
-        header("HTTP/1.0 204 No Content");
-        exit;
+        // uploaded file didn't exist or filesize didn't match, request a re-upload
+        http_response_code(204);
     }
 }
+// Handle a file upload request
+elseif ($_SERVER['REQUEST_METHOD'] === 'POST') {
+    foreach ($_FILES as $file) {
+        if ($file["error"] != UPLOAD_ERR_OK) {
+            report_error(get_upload_err_msg($file['error']));
+            exit;
+        }
 
-// handle a file upload request
-foreach ($_FILES as $file) {
-    if ($file["error"] != UPLOAD_ERR_OK) {
-        report_error(get_upload_err_msg($file['error']), false);
-        exit;
-    }
+        if (!is_dir($staging_dir)) {
+            // Suppress warning by mkdir because despite the check above
+            // race conditions from other processes can result in the
+            // directory existing when we try to create it here.
+            @mkdir($staging_dir, 0777, true);
+        }
 
-    if (!is_dir($staging_dir)) {
-        // Suppress warning by mkdir because despite the check above
-        // race conditions from other processes can result in the
-        // directory existing when we try to create it here.
-        @mkdir($staging_dir, 0777, true);
+        if (!move_uploaded_file($file['tmp_name'], $chunk_filename)) {
+            report_error("Error saving chunk $chunk_filename for $filename");
+            exit;
+        } elseif (filesize($chunk_filename) < $chunk_size) {
```

Comment:
I think it is this line that is causing the problem I am seeing. If the whole file size is less than the chunk size it will always fail here.

Reply:
Yup. I was able to reproduce this too. Pushed up a new commit to handle this case.